tests/test_core.py

```diff
@@ -0,0 +1,111 @@
+import pathlib
+import platform
+from operator import attrgetter
+
+import pytest  # type: ignore
+
+if platform.python_implementation() == "PyPy":
+    from yaml import load, SafeLoader
+else:
+    try:
+        from yaml import load, CSafeLoader as SafeLoader  # type: ignore
+    except ImportError:
+        logging.getLogger(__name__).warning(
+            "PyYaml C extension not available to run tests, this will result "
+            "in dramatic tests slowdown."
+        )
+        from yaml import load, SafeLoader
+
+from ua_parser import *
+
+CORE_DIR = (pathlib.Path(__name__).parent.parent / "uap-core").resolve()
+
+
+PARSERS = [
+    pytest.param(BasicParser.from_regexes(), id="compiled"),
+    pytest.param(BasicParser.from_yaml(CORE_DIR / "regexes.yaml"), id="yaml"),
+    pytest.param(
+        CachingParser(
+            BasicParser.from_regexes(),
+            Clearing(100),
+        ),
+        id="clearing",
+    ),
+    pytest.param(
+        CachingParser(
+            BasicParser.from_regexes(),
+            LRU(100),
+        ),
+        id="lru",
+    ),
+]
+
+
+UA_FIELDS = {f.name for f in dataclasses.fields(UserAgent)}
+
+
+@pytest.mark.parametrize("parser", PARSERS)
+@pytest.mark.parametrize(
+    "test_file",
+    [
+        CORE_DIR / "tests" / "test_ua.yaml",
+        CORE_DIR / "test_resources" / "firefox_user_agent_strings.yaml",
+        CORE_DIR / "test_resources" / "pgts_browser_list.yaml",
+    ],
+    ids=attrgetter("name"),
+)
+def test_ua(parser, test_file):
+    with test_file.open("rb") as f:
+        contents = load(f, Loader=SafeLoader)
+
+    for test_case in contents["test_cases"]:
+        res = {k: v for k, v in test_case.items() if k in UA_FIELDS}
+        r = parser.parse_user_agent(test_case["user_agent_string"])
+        assert dataclasses.asdict(r) == res
+
+
+OS_FIELDS = {f.name for f in dataclasses.fields(OS)}
+
+
+@pytest.mark.parametrize("parser", PARSERS)
+@pytest.mark.parametrize(
+    "test_file",
+    [
+        CORE_DIR / "tests" / "test_os.yaml",
+        CORE_DIR / "test_resources" / "additional_os_tests.yaml",
+    ],
+    ids=attrgetter("name"),
+)
+def test_os(parser, test_file):
+    with test_file.open("rb") as f:
+        contents = load(f, Loader=SafeLoader)
+
+    for test_case in contents["test_cases"]:
+        res = {k: v for k, v in test_case.items() if k in OS_FIELDS}
+        r = parser.parse_os(test_case["user_agent_string"])
+        assert dataclasses.asdict(r) == res
+
+
+DEVICE_FIELDS = {f.name for f in dataclasses.fields(Device)}
+
+
+@pytest.mark.parametrize("parser", PARSERS)
+@pytest.mark.parametrize(
+    "test_file",
+    [
+        CORE_DIR / "tests" / "test_device.yaml",
+    ],
+    ids=attrgetter("name"),
+)
+def test_devices(parser, test_file):
+    with test_file.open("rb") as f:
+        contents = load(f, Loader=SafeLoader)
+
+    for test_case in contents["test_cases"]:
+        res = {k: v for k, v in test_case.items() if k in DEVICE_FIELDS}
```

Comment:
Maybe just implement #142 instead of bothering with this filtering (which mirrors the legacy behaviour: https://github.com/ua-parser/uap-python/blob/master/tests/test_legacy.py#L174-L178)